Remotion/Web/Core/Context/HttpContextStorageProviderNetCore.cs

```diff
@@ -0,0 +1,208 @@
+        {
+          Context = context;
+          Data = data;
+        }
+      }
+
+      private const string c_dataDictionaryKey = "Rm.DD";
+
+      private readonly ThreadLocal<WeakReference<WrappedHashtable?>> _dataTable;
+
+      public HashTableAccessor (bool _)
+      {
+        _dataTable = new(() => new WeakReference<WrappedHashtable?>(null));
+      }
+
+      public bool TryGetCachedHashtable (HttpContext httpContext, [MaybeNullWhen(false)] out Hashtable hashtable)
+      {
+        if (TryGetCachedWrappedHashtable(httpContext, out var cachedWrappedHashtable))
+        {
+          hashtable = cachedWrappedHashtable.Data;
+          return true;
+        }
+
+        hashtable = null;
+        return false;
+      }
+
+      public Hashtable GetOrCreateHashtable (HttpContext context)
+      {
+        if (TryGetCachedHashtable(context, out var hashtable))
+        {
+          return hashtable;
+        }
+
+        return SetHashtable(context, new Hashtable());
+      }
+
+      public Hashtable? RemoveHashtable (HttpContext context)
+      {
+        if (TryGetCachedWrappedHashtable(context, out var cachedHashtable))
+        {
+          context.Items.Remove(c_dataDictionaryKey);
+          _dataTable.Value!.SetTarget(null!);
+        }
+
+        return cachedHashtable?.Data;
+      }
+
+      public Hashtable SetHashtable (HttpContext context, Hashtable hashtable)
+      {
+        var wrappedHashtable = new WrappedHashtable(context, hashtable);
+        context.Items[c_dataDictionaryKey] = wrappedHashtable;
+        _dataTable.Value!.SetTarget(wrappedHashtable);
+        return hashtable;
+      }
+
+      private bool TryGetCachedWrappedHashtable (HttpContext httpContext, [MaybeNullWhen(false)] out WrappedHashtable hashtable)
+      {
+        if (_dataTable.Value!.TryGetTarget(out var target) && target.Context == httpContext)
+        {
+          hashtable = target;
+          return true;
+        }
+
+        var wrappedHashtable = (WrappedHashtable?)httpContext.Items[c_dataDictionaryKey];
+        hashtable = wrappedHashtable;
+
+        if (wrappedHashtable != null)
+        {
+          _dataTable.Value!.SetTarget(wrappedHashtable);
+#pragma warning disable CS8762 // Parameter must have a non-null value when exiting in some condition.
+          return true;
+#pragma warning restore CS8762 // Parameter must have a non-null value when exiting in some condition.
+        }
+
+        return false;
+      }
+    }
+
+    // Ignore the obsolete warnings for AsyncLocalStorageProvider
+#pragma warning disable 618
+    private readonly AsyncLocalStorageProvider _fallbackProvider = new();
+#pragma warning restore 618
+
+    private readonly HashTableAccessor _hashTableAccessor = new(true);
+
+    public SafeContextBoundary OpenSafeContextBoundary ()
+    {
+      var httpContext = HttpContext.Current;
+      if (httpContext == null)
+      {
+        return _fallbackProvider.OpenSafeContextBoundary();
+      }
```

Comment:
Style remark: We are mixing the way we handle the storage fallback quite a lot. Here we check for the fallback first, in the other methods we do it second. We also have `else` statements below while an exit `if` here. Maybe we should unify this to have the same pattern everywhere?

My personal preference would be to do the fallback right after the argument check like so:

```c#
// arg checks
if (httpContext == null)
  return _fallbackProvider.XXXX();
// rest
```